Userland/Applications/PixelPaint/Tools/GradientTool.cpp

```diff
@@ -0,0 +1,481 @@
+        opacity_slider.set_range(1, 100);
+        opacity_slider.set_value(100);
+
+        opacity_slider.on_change = [&](int value) {
+            m_opacity = value;
+            m_editor->update();
+        };
+
+        set_primary_slider(&opacity_slider);
+
+        auto& button_container = m_properties_widget->add<GUI::Widget>();
+        button_container.set_fixed_height(22);
+        auto& button_container_layout = button_container.set_layout<GUI::HorizontalBoxLayout>();
+        button_container_layout.add_spacer();
+
+        auto& apply_button = button_container.add<GUI::DialogButton>("Apply");
+        apply_button.on_click = [this](auto) {
+            rasterize_gradient();
+        };
+    }
+
+    return m_properties_widget.ptr();
+}
+
+void GradientTool::rasterize_gradient()
+{
+    if (!has_gradient_start_end())
+        return;
+
+    GUI::Painter painter(m_editor->active_layer()->get_scratch_edited_bitmap());
+    draw_gradient(painter);
+
+    m_editor->did_complete_action("Gradient Tool"sv);
+
+    reset();
+}
+
+void GradientTool::calculate_gradient_lines()
+{
+    m_gradient_half_length = m_gradient_end.value().distance_from(m_gradient_center.value());
+
+    // create a perpendicular point between center and end
+    m_perpendicular_point = m_gradient_end.value();
+    m_perpendicular_point -= m_gradient_center.value();
+    m_perpendicular_point.transform_by(m_transform_perpendicular);
+    m_perpendicular_point += m_gradient_center.value();
+
+    auto to_edge_scale_direction = (m_physical_diagonal_layer_length * 2) / m_gradient_center.value().distance_from(m_perpendicular_point);
+
+    m_gradient_center_line.set_a({ m_gradient_center.value().x() + (to_edge_scale_direction * (m_gradient_center.value().x() - m_perpendicular_point.x())), m_gradient_center.value().y() + (to_edge_scale_direction * (m_gradient_center.value().y() - m_perpendicular_point.y())) });
+    m_gradient_center_line.set_b({ m_gradient_center.value().x() + (-to_edge_scale_direction * (m_gradient_center.value().x() - m_perpendicular_point.x())), m_gradient_center.value().y() + (-to_edge_scale_direction * (m_gradient_center.value().y() - m_perpendicular_point.y())) });
+
+    m_gradient_begin_line.set_a(m_gradient_center_line.a().translated(static_cast<Gfx::FloatPoint>(m_gradient_end.value() - m_gradient_center.value())));
+    m_gradient_begin_line.set_b(m_gradient_center_line.b().translated(static_cast<Gfx::FloatPoint>(m_gradient_end.value() - m_gradient_center.value())));
+
+    m_gradient_end_line.set_a(m_gradient_center_line.a().translated(static_cast<Gfx::FloatPoint>(m_gradient_center.value() - m_gradient_end.value())));
+    m_gradient_end_line.set_b(m_gradient_center_line.b().translated(static_cast<Gfx::FloatPoint>(m_gradient_center.value() - m_gradient_end.value())));
+    m_editor->update();
+}
+
+void GradientTool::draw_gradient(GUI::Painter& painter, bool with_guidelines, const Gfx::FloatPoint drawing_offset, float scale)
```

Comment:
(Sorry don't mean to sound mean :sweat_smile: I've just spend a long time looking at gradients in LibWeb)

This seems like a very overly complex way to draw a gradient and I don't really understand why?

Simply put to draw a gradient you just need to compute a 1D gradient line.
This is simply an array of colors the length of your gradient, all nicely blended with smooth transitions.
Then for you gradient region (i.e. the rectangle bounding box), you take each (x, y) coordinate and apply some a mapping from that point to the gradient line. This mapping can be whatever you like, e.g. if it returned distance from the center you get a radial gradient, with a little rotation of the point you can do a linear gradient (of any angle), or a conic gradient etc, etc. This how I paint gradients in LibWeb and the resulting implementing (to me anyway), is quite a bit simpler and easier to understand (and reasonably efficient since you do all the blending only on the 1D gradient line).

My thoughs here are perhaps I should move my gradient painting code over to LibGfx, which things like this tool and other parts of the system could use? Because I think CSS gradient rendering pretty much covers every effect you could want.

Reply:
I've integrated the new gradient api, so the code looks much cleaner now. Furthermore the clipping has been optimized when zoomed in and the handles are easier to use when zoomed out.

Still open issues where i currently have no solution:
- visual glitches while there is a in-progress gradient and the mouse is moved arround - (this can cause double redraws)
- applying gradient with a selection still does nothing